src/AElf.Runtime.WebAssembly/ExternalEnvironment.cs

```diff
@@ -0,0 +1,231 @@
+    {
+        throw new NotImplementedException();
+    }
+
+    public async Task<byte[]?> GetStorageAsync(byte[] key)
+    {
+        var stateKey = GetStateKey(key);
+        if (Writes.ContainsKey(stateKey))
+        {
+            if (Writes.TryGetValue(stateKey, out var byteStringValue))
+            {
+                Reads.TryAdd(stateKey, true);
+                return byteStringValue.ToByteArray();
+            }
+        }
+
+        var value = await HostSmartContractBridgeContext!.GetStateAsync(stateKey);
+        var byteArrayValue = value?.ToByteArray();
+        Reads.TryAdd(stateKey, value != null);
+        return byteArrayValue;
+    }
+
+    public async Task<int> GetStorageSizeAsync(byte[] key)
+    {
+        var value = await GetStorageAsync(key);
+        return value?.Length ?? 0;
+    }
+
+    public bool IsContract(byte[] address)
+    {
+        throw new NotImplementedException();
+    }
+
+    public Hash? CodeHash(byte[] address)
+    {
+        throw new NotImplementedException();
+    }
+
+    public Hash OwnCodeHash()
+    {
+        throw new NotImplementedException();
+    }
+
+    public bool CallerIsOrigin()
+    {
+        throw new NotImplementedException();
+    }
+
+    public bool CallerIsRoot()
+    {
+        throw new NotImplementedException();
+    }
+
+    public Address Address()
+    {
+        throw new NotImplementedException();
+    }
+
+    public long Balance()
+    {
+        throw new NotImplementedException();
+    }
+
+    public long GetWeightPrice(Weight weight)
+    {
+        throw new NotImplementedException();
+    }
+
+    public long ValueTransferred()
+    {
+        throw new NotImplementedException();
+    }
+
+    public byte[]? EcdsaRecover(byte[] signature, byte[] messageHash)
+    {
+        CryptoHelper.RecoverPublicKey(signature, messageHash, out var pubkey);
+        return pubkey;
+    }
+
+    public long Now()
+    {
+        return TimestampHelper.GetUtcNow().Seconds;
+    }
+
+    public long MinimumBalance()
+    {
+        throw new NotImplementedException();
+    }
+
+    public (byte[], long) Random(byte[] subject)
+    {
+        throw new NotImplementedException();
+    }
+
+    public void DepositEvent(byte[] topics, byte[] data)
+    {
+        throw new NotImplementedException();
+    }
+
+    public long BlockNumber()
+    {
+        throw new NotImplementedException();
+    }
+
+    public int MaxValueSize()
+    {
+        throw new NotImplementedException();
+    }
+
+    public bool AppendDebugBuffer(string message)
+    {
+        throw new NotImplementedException();
+    }
+
+    public byte[] EcdsaToEthAddress(byte[] pubkey)
+    {
+        var pubkeyNoPrefixCompressed = new byte[pubkey.Length - 1];
+        Array.Copy(pubkey, 1, pubkeyNoPrefixCompressed, 0, pubkeyNoPrefixCompressed.Length);
+        var initAddress = new Sha3Keccack().CalculateHash(pubkeyNoPrefixCompressed);
+        var address = new byte[initAddress.Length - 12];
```

Comment:
similarly just to check can we assume initAddress will always be >12?

Reply:
Checked the impl of new Sha3Keccack().CalculateHash, yeah it's true that initAddress will always be > 12